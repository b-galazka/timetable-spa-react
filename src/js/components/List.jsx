import React, {Component} from 'react';
import {connect} from 'react-redux';
import {Link, withRouter} from 'react-router-dom';

function mapStateToProps(state) {

    return {
        teachers: state.teachers,
        classes: state.classes,
        classrooms: state.classrooms
    };
}

class List extends Component {

    render() {

        return (
            <ul className="sidebar__list">
                {this.renderItems()}
            </ul>
        );
    }

    renderItems() {

        const type = this.props.type;

        switch (type) {

            case 'teacher':
                return this.renderTeachers();

            case 'class':
                return this.renderClasses();

            case 'classroom':
                return this.renderClassrooms();
        }
    }

    renderTeachers() {

        const teachers = this.props.teachers;

        const teacherUrlParam = decodeURIComponent(this.props.match.params.slug);

        return teachers.map((teacher) => {

            const url = `/nauczyciel/${encodeURIComponent(teacher.slug)}`;

            const link = (
                <Link to={url}>
                    {teacher.name || teacher.slug}
                </Link>
            );

            return (
                teacherUrlParam === teacher.slug ?

                <li key={teacher._id} className="active">{link}</li> :

                <li key={teacher._id}>{link}</li>
            );    
        });
    }

    renderClasses() {

        const classes = this.props.classes;

        const classUrlParam = decodeURIComponent(this.props.match.params.slug);

        return classes.map((schoolClass) => {

            const url = `/klasa/${encodeURIComponent(schoolClass.slug)}`;

            const link = (
                <Link to={url}>
                    {schoolClass.slug}
                </Link>
            );

            return (
                classUrlParam === schoolClass.slug ?

                <li key={schoolClass._id} className="active">{link}</li> :

                <li key={schoolClass._id}>{link}</li>
            );
        });
    }

    renderClassrooms() {

        const classrooms = this.props.classrooms;

        const classroomUrlParam = decodeURIComponent(this.props.match.params.slug);

        return classrooms.map((classroom) => {

            const url = `/sala/${encodeURIComponent(classroom.number)}`;

            const link = (
                <Link to={url}>
                    {classroom.number}
                </Link>
            );

            return (
                classroomUrlParam === classroom.number ?

                <li key={classroom._id} className="active">{link}</li> :

                <li key={classroom._id}>{link}</li>
            );
        });
    }
}

export default withRouter(connect(mapStateToProps)(List));