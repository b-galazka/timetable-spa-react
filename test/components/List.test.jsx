import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import List from '../../src/js/components/List';

Enzyme.configure({ adapter: new Adapter() });

jest.mock(
    '../../src/json/urlTranslations',
    () => ({
        teacher: 'teacher',
        class: 'class',
        classroom: 'classroom'
    })
);

describe('List component', () => {

    it('should render teachers list', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <List.WrappedComponent
                    onLinkClicked={jest.fn()}
                    type="teacher"
                    classrooms={[]}
                    classes={[]}

                    match={{
                        params: { slug: 'XYZ', type: 'teacher' }
                    }}

                    teachers={[

                        {
                            slug: 'ABC',
                            _id: "1",
                            name: null
                        },

                        {
                            slug: 'XYZ',
                            _id: "2",
                            name: 'XYZ name'
                        },

                        {
                            slug: '#ZZZ',
                            _id: "3",
                            name: 'ZZZ name'
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render classes list', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <List.WrappedComponent
                    onLinkClicked={jest.fn()}
                    type="class"
                    classrooms={[]}
                    teachers={[]}

                    match={{
                        params: { slug: '4ET', type: 'class' }
                    }}

                    classes={[

                        {
                            slug: '1AL',
                            _id: "1"
                        },

                        {
                            slug: '4ET',
                            _id: "2"
                        },

                        {
                            slug: '#3ET',
                            _id: "3"
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render classrooms list', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <List.WrappedComponent
                    onLinkClicked={jest.fn()}
                    type="classroom"
                    classes={[]}
                    teachers={[]}

                    match={{
                        params: { slug: '200', type: 'classroom' }
                    }}

                    classrooms={[

                        {
                            number: '100',
                            _id: "1"
                        },

                        {
                            number: '200',
                            _id: "2"
                        },

                        {
                            number: '#300',
                            _id: "3"
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call provided props.onLinkClicked function on link click', () => {

        const onLinkClickedMock = jest.fn();

        const list = mount(
            <MemoryRouter>
                <List.WrappedComponent
                    onLinkClicked={onLinkClickedMock}
                    type="teacher"
                    classes={[]}
                    classrooms={[]}

                    match={{
                        params: { slug: '200', type: 'teacher' }
                    }}

                    teachers={[

                        {
                            slug: 'XYZ',
                            _id: "2",
                            name: 'XYZ name'
                        }
                    ]}
                />
            </MemoryRouter>
        );

        list.find('li a').simulate('click');

        expect(onLinkClickedMock).toHaveBeenCalled();
    });
});