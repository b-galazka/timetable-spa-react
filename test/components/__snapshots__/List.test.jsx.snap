// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render classes list 1`] = `
<ul
  className="sidebar__list"
>
  <li>
    <a
      href="/class/1AL"
      onClick={[Function]}
    >
      1AL
    </a>
  </li>
  <li
    className="active"
  >
    <a
      href="/class/4ET"
      onClick={[Function]}
    >
      4ET
    </a>
  </li>
  <li>
    <a
      href="/class/%233ET"
      onClick={[Function]}
    >
      #3ET
    </a>
  </li>
</ul>
`;

exports[`List component should render classrooms list 1`] = `
<ul
  className="sidebar__list"
>
  <li>
    <a
      href="/classroom/100"
      onClick={[Function]}
    >
      100
    </a>
  </li>
  <li
    className="active"
  >
    <a
      href="/classroom/200"
      onClick={[Function]}
    >
      200
    </a>
  </li>
  <li>
    <a
      href="/classroom/%23300"
      onClick={[Function]}
    >
      #300
    </a>
  </li>
</ul>
`;

exports[`List component should render teachers list 1`] = `
<ul
  className="sidebar__list"
>
  <li>
    <a
      href="/teacher/ABC"
      onClick={[Function]}
    >
      ABC
    </a>
  </li>
  <li
    className="active"
  >
    <a
      href="/teacher/XYZ"
      onClick={[Function]}
    >
      XYZ name
    </a>
  </li>
  <li>
    <a
      href="/teacher/%23ZZZ"
      onClick={[Function]}
    >
      ZZZ name
    </a>
  </li>
</ul>
`;
